### ISSUE TITLE
Write down the first version CONTRIBUTING.md

### ISSUE THREAD

**Issue by tisonkun** [engula/engula#19]
In order not to disturb your prototyping and rapidly development, I'm glad to write this file and propose a PR, with your help on collecting necessary information.

The draft is looks like:

---

# How to Contribute

I'm really glad you're reading this, because we need volunteer developers to help this project come to fruition.

If you haven't already, come [find us on gitter](https://gitter.im/engula/contributors?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge). We want you working on things you're excited about.

Welcome to review [our design](https://github.com/engula/engula/blob/main/docs/design.md) or participant discussions about [the roadmap](https://github.com/engula/engula/issues/1)!

## Get Started

We develop Engula with rust stable toolchain.

You're able to get started with Engula with three steps:

1. Setup the environment with [rustup](https://rustup.rs/).
2. Build Engula via `cargo build`.
3. Run the example via `cargo run --example hello`.

## Report an Issue

If you think you have found an issue in Engula, you can report it to the [issue tracker](https://github.com/engula/engula/issues).

Before filing an issue report is to see whether the problem has already been reported. You can [use the search bar to search existing issues](https://docs.github.com/en/github/administering-a-repository/finding-information-in-a-repository/using-search-to-filter-issues-and-pull-requests). This doesn't always work, and sometimes it's hard to know what to search for, so consider this extra credit. We won't mind if you accidentally file a duplicate report. Don't blame yourself if your issue is closed as duplicated.

If the problem you're reporting is not already in the issue tracker, you can [open a GitHub issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-an-issue) with your GitHub account.

## Submitting a Pull Request

Please send a [GitHub Pull Request to Engula](https://github.com/engula/engula/pull/new/main) with a clear list of what you've done (read more about [pull requests](http://help.github.com/pull-requests/)). When you send a pull request, we're looking forward to an expressive description, clear commit messages, and more test coverage if it is code contribution.

Before submitting the pull request, please make sure all tests pass locally:

```bash
cargo build --release
cargo test
cargo clippy -- -D warnings
cargo fmt --all -- --check
```

Thank you for your participation!

---

The questions are:

1. Do we protect `main` and only modify it by PR?
2. What merge strategy do we use, specially, merge with commit, rebase and merge, or squash and merge? I highly recommend the latter two where merge with commit make history hard to read - however, I'm not participant your project deeply, so it's your choice.
3. Shall we adopt a code of conduct? If so, I'd suggest [Contributor Covenant Code of Conduct](https://www.contributor-covenant.org/version/2/1/code_of_conduct/code_of_conduct.md) and the project should provide a contact method.
4. Any other concern on the draft above?

**Comment by huachaohuang** [comment]
Wow, there are a lot of things here. Give me some time to figure them out :)

**Comment by tisonkun** [comment]
Hi @huachaohuang ! Any progress on this issue? I'm waiting for a conclusion on whether this is good to go.

**Comment by huachaohuang** [comment]
Sorry, I'm still busy with the demo. The whole demo stage will finish at the end of Nov 2021. I will initiate Community-related work in Dec 2021. So I may not work on this issue until then.

**Comment by tisonkun** [comment]
@huachaohuang do you think you guys should take another consideration on this proposal?

I don't mean to throw a whole "community plan" or something, but the specific CONTRIBUTING file and an optional CODE_OF_CONDUCT file.

You can grow your bandwidth on multiple domains by shepherding the direction but delegating works to the whole community.

**Comment by huachaohuang** [comment]
Surely I can be time-sharing. So the problem is not that I really have no time to review these files. They are just some common words that do not harm anyone. But adding these files without a clear goal is not how we work. I don't do it simply because everyone else does it.
On the other hand, the primary goal in this stage is not about making rules but about where we go. People are not going to board our ship unless they share the same direction with us.

**Comment by tisonkun** [comment]
@huachaohuang that makes sense. Thanks for your time and hope your demo goes well.

Of course we should try to be clear with the direction where we go, while the community shape will have its own prototype stage :)

**Comment by tisonkun** [comment]
@huachaohuang I agree that you should be aware of which Code of Conduct you choose because we will follow them and exercise them ourselves.

For the CONTRIBUTING.md part, you can regard it as a quickstart of developer and require less consideration. They are not unclear "common words that do not harm anyone" - they are a clear quickstart of contributors.

**Comment by huachaohuang** [comment]
Sure, I share the same view with you. I will consider that seriously when the time comes :)
